### PR TITLE
Spark: Use options as a CaseSensitiveMap

### PR DESCRIPTION
**Pull request by rajarshisarkar**
Recently, it was reported in Slack general channel that [AWS S3 tagging](https://github.com/apache/iceberg/blob/master/docs/integrations/aws.md#s3-tags) present in master branch is not working for keys with upper case.

Reference: https://apache-iceberg.slack.com/archives/C025PH0G1D4/p1653879105452189

This is because we use `CaseInsensitiveStringMap` for options in `SparkCatalog`. This PR uses options as a `CaseSensitiveMap`.

With this change, the following Spark configurations:
```
sh spark-sql --conf spark.sql.catalog.my_catalog=org.apache.iceberg.spark.SparkCatalog \
    --conf spark.sql.catalog.my_catalog.warehouse=s3://iceberg-dev/warehouse.case-sensitive \
    --conf spark.sql.catalog.my_catalog.catalog-impl=org.apache.iceberg.aws.glue.GlueCatalog \
    --conf spark.sql.catalog.my_catalog.io-impl=org.apache.iceberg.aws.s3.S3FileIO \
    --conf spark.sql.catalog.my_catalog.s3.write.tags.my_key=my_val \
    --conf spark.sql.catalog.my_catalog.s3.write.tags.MY_key2=my_val2
```
now saves the S3 tags as:

```
my_key = my_val
MY_key2 = my_val2
```

---

cc: @RussellSpitzer @jackye1995

### CODE REVIEW

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkCatalog.java [hunk 1]
```diff
@@ -119,7 +119,7 @@ public class SparkCatalog extends BaseCatalog {
   protected Catalog buildIcebergCatalog(String name, CaseInsensitiveStringMap options) {
     Configuration conf = SparkUtil.hadoopConfCatalogOverrides(SparkSession.active(), name);
     Map<String, String> optionsMap = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
-    optionsMap.putAll(options);
+    optionsMap.putAll(options.asCaseSensitiveMap());
```
Comment (rdblue) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkCatalog.java:122]: Why is this needed? The `optionsMap` is a treemap that is case insensitive.

Comment (rajarshisarkar) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkCatalog.java:122]: This is needed to save case sensitive [S3 write/delete tag keys](https://github.com/apache/iceberg/blob/master/docs/integrations/aws.md#s3-tags) for objects.

There is a thread in `general` channel on this, let me add you there: https://apache-iceberg.slack.com/archives/C025PH0G1D4/p1653879105452189

I have removed the case insensitive comparator from the Treemap as well.

Comment (rdblue) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkCatalog.java]: Got it. So this passes the original case for when we extract the keys. Thanks!

Comment (rajarshisarkar) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkCatalog.java]: Yes, correct! 

Thanks, for merging. I'll backport the changes for Spark 3.x.

Comment (rajarshisarkar) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkCatalog.java]: I have raised the backport PR for Spark 3.x: https://github.com/apache/iceberg/pull/5044

**Comment by rdblue** [comment]
@rajarshisarkar, I just reverted this in master because it broke the contract and made the options map case sensitive. Can you implement my comment on #5044 to get this in?

**Comment by rajarshisarkar** [comment]
Sure, @rdblue. I have updated https://github.com/apache/iceberg/pull/5044.
